### ISSUE TITLE
More multilingual, more better?

### ISSUE THREAD

**Issue by tom9358** [tom9358/nllb-tryout#35]
Previously, I saw substantial improvement of the NLLB models when adding English+German+Spanish multilingual data. We can now add languages from Tatoeba without increasing the total amount of data, with the `focus_total` argument:

```python
cfg = RunConfig(
    source_langs_tatoeba=("nld", "gos", "eng", "deu", "spa", "fra"),
    source_langs_nllb=("nld_Latn", "gos_Latn", "eng_Latn", "deu_Latn", "spa_Latn", "fra_Latn"),
    sampling_strategy="focus_total",
    focus_lang_pair=("nld_Latn", "gos_Latn"),
    sampling_temperature=5.0,
)
```

This means it is now feasible to investigate the effect of adding/leaving out any Tatoeba language (that's also supported by NLLB; I think the code still doesn't support having more than one new language added to the tokenizer).

I wonder if there is a point of diminishing returns. Let's work out this issue later in more detail to make the DoD / research question(s) more concrete.

**Comment by tom9358** [comment]
<details><summary>fun to read related LLM barf</summary>
Title:

feat: test French in fixed-budget multilingual fine-tuning

Body:

## Background

The project fine-tunes NLLB for translation between Dutch and Gronings. The original training data consisted primarily of approximately 13,000 clean Dutch–Gronings sentence pairs from Tatoeba.

In #34, a much larger synthetic corpus was generated by translating authentic Gronings Kreuze text into Dutch with Gemma. Controlled experiments showed that this synthetic data is useful, especially when training is split into two stages:

1. pooled multilingual training on Tatoeba and synthetic Kreuze data;
2. a short clean multilingual Tatoeba finish.

The best tested model so far uses NLLB 1.3B and:

- Dutch, Gronings, English, German and Spanish;
- eight pooled multilingual epochs;
- 230,712 samples per epoch;
- 50% Dutch–Gronings samples;
- 50% supporting-language samples;
- temperature-balanced supporting pairs with `T=5`;
- a final clean Tatoeba pass containing:
  - all 13,038 clean Dutch–Gronings training pairs;
  - 13,038 temperature-balanced supporting-language samples.

This reached approximately 70.79 BLEU / 85.30 chrF on the complete Dutch→Gronings Tatoeba validation set, while retaining large gains on the independent Kreuze validation set.

Longer multilingual training also largely eliminated the supporting-language regression seen in shorter synthetic-data experiments. At eight epochs, synthetic-plus-clean versus the equal-compute Tatoeba-only control differed by only -0.33 macro BLEU / -0.02 macro chrF across the other 18 translation directions.

Full experiment details are documented in [`data/kreuze/training_plan.md`](https://github.com/tom9358/nllb-tryout/blob/main/data/kreuze/training_plan.md).

## Question

Would adding French improve Dutch–Gronings translation through multilingual transfer, or would it merely dilute the fixed supporting-language budget?

French is an interesting candidate because:

- NLLB already has strong pretrained French support;
- Tatoeba contains approximately 1,000 French–Gronings pairs;
- French may add useful syntactic and lexical variation;
- the fixed-budget `focus_total` sampler now makes adding languages computationally feasible.

However, adding French is not free. Five languages produce ten unordered language pairs, or twenty directed translation directions. Six languages produce fifteen unordered pairs, or thirty directed directions.

Excluding the primary Dutch↔Gronings pair, the supporting mixture would grow from:

- 9 unordered pairs / 18 directions;
- to 14 unordered pairs / 28 directions.

The same 115,356-sample supporting budget would therefore be distributed across more corpora. Existing English, German and Spanish pairs would receive fewer samples unless the allocation strategy is changed.

## Proposed experiment

Compare otherwise identical five-language and six-language runs.

### Five-language condition

- Dutch
- Gronings
- English
- German
- Spanish

### Six-language condition

- Dutch
- Gronings
- English
- German
- Spanish
- French

Both conditions should use:

- NLLB 1.3B;
- the same seed;
- the same total samples per epoch;
- the same optimizer-step budget;
- the same batch size and maximum sequence length;
- the same alternating-direction strategy;
- the same synthetic Kreuze corpus;
- the same 50:50 Dutch–Gronings/supporting split;
- temperature balancing at `T=5` within the supporting half;
- the same clean-finish recipe;
- no RUG data.

The exact per-corpus sample allocation should be recorded before training.

## Validation-split requirement

The current global Tatoeba hold-out depends on the configured language set. Simply adding French changes the sentence-ID universe and therefore changes the validation split.

The five- and six-language models must not be compared using different Dutch–Gronings validation sets.

Use a common frozen sentence-ID hold-out for both conditions—for example, construct the hold-out from the union of all six languages and use that exact assignment for both runs. This probably means rerunning the five-language baseline rather than directly comparing against the existing checkpoint.

## Suggested stages

### 1. Short screening experiment

Run a matched short-budget comparison first, using the same number of base epochs and the same clean finish for both language sets.

Stop if French clearly:

- lowers Dutch↔Gronings validation performance;
- causes substantial regressions across the existing 18 supporting directions;
- or provides little useful French↔Gronings capability.

### 2. Full-budget experiment

If the short comparison is promising, repeat the best #34 recipe:

- eight pooled multilingual epochs;
- followed by the chosen clean-finish length.

## Evaluation

Report held-out validation scores only, clearly separated from training diagnostics.

### Primary metrics

Complete frozen Tatoeba validation:

- Dutch→Gronings BLEU and chrF;
- Gronings→Dutch BLEU and chrF.

Independent Kreuze validation:

- Dutch→Gronings BLEU and chrF;
- Gronings→Dutch BLEU and chrF.

### Existing-language regression checks

Report all 18 existing supporting directions and their macro averages:

- Dutch↔English
- Dutch↔German
- Dutch↔Spanish
- Gronings↔English
- Gronings↔German
- Gronings↔Spanish
- English↔German
- English↔Spanish
- German↔Spanish

### French metrics

Report all ten new French directions:

- French↔Dutch
- French↔Gronings
- French↔English
- French↔German
- French↔Spanish

Also report:

- macro BLEU and chrF across the ten French directions;
- macro BLEU and chrF across the original 18 supporting directions;
- macro BLEU and chrF across all 28 supporting directions.

If the Dutch↔Gronings difference is small, repeat with additional seeds before drawing a strong conclusion.

## Questions to answer

- Does adding French improve Dutch↔Gronings generalization?
- Does it improve Gronings↔French enough to be practically useful?
- Does dividing the same supporting budget among fourteen rather than nine pairs damage the existing language directions?
- Does temperature balancing give the relatively small French–Gronings corpus enough exposure?
- Is six-language training worth the additional complexity while keeping total compute fixed?

## Scope

This issue is about multilingual transfer and language-set composition. It should not reopen the synthetic-generation investigation from #34 or increase the total training budget merely because another language was added.
</details>
